### ISSUE TITLE
RaymondFam - Later borrowers will pay less interest compared to earlier borrowers over the same period

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-04-jojo-judging#246]
RaymondFam

medium

# Later borrowers will pay less interest compared to earlier borrowers over the same period

## Summary
The way the interest calculation is designed, later borrowers will pay less interest compared to earlier borrowers over the same period, assuming `borrowFeeRate` remains unchanged.

## Vulnerability Detail
Suppose the first user borrowed JUSD at t0, the total debt plus interest say 10% per year incurred in 6 months is amount borrowed * 1.05/1, i.e. interest is 5%. If someone were to borrow at this time, the interest incurred 6 months later would be (1.1/1.05 - 1)*100% = 4.76%. This will be more and more significant down the road. For instance, borrowing 10 years later will have the interest (2/1.95 - 1)*100% = 2.56%. All future borrowers are paying less and lesser interest. 

## Impact
This could be seen as unfair to earlier borrowers and might not be the desired behavior for your system.

## Code Snippet
https://github.com/sherlock-audit/2023-04-jojo/blob/main/JUSDV1/src/Impl/JUSDBankStorage.sol#L51-L57

## Tool used

Manual Review

## Recommendation
One way to address this issue is to modify the interest calculation method to calculate interest rates based on the time each borrower took the loan. This can be done by storing the timestamp for each borrower when they take a loan and using it in the interest calculation.

For instance, you can create a new mapping to store the borrowing timestamp for each borrower:

```solidity
mapping(address => uint256) public borrowingTimestamps;
```
When a user borrows, update the timestamp:
```solidity
borrowingTimestamps[borrower] = block.timestamp;
```
And when calculating the interest rate for repayment, use the borrower's borrowing timestamp instead of the global `lastUpdateTimestamp`:
```solidity
uint256 timeDifference = block.timestamp - borrowingTimestamps[borrower];
```
This way, each borrower's interest rate will be calculated independently based on the time elapsed since they borrowed, ensuring fairness across all users.

Keep in mind that this approach will increase the storage requirements for your contract, as you'll need to store individual timestamps for each borrower. Additionally, it will require refactoring in other related codes. However, it ensures that each borrower pays interest based on their borrowing time, which can be considered a fairer approach.

Duplicate of #73

**Comment by 0xruhum** [comment]
Escalate for 10 USDC

This describes the same issue as #73. The borrow rate calculation is broken causing it to be less than expected.

**Comment by sherlock-admin** [comment]
> Escalate for 10 USDC
> 
> This describes the same issue as #73. The borrow rate calculation is broken causing it to be less than expected.

You've created a valid escalation for 10 USDC!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by Trumpero** [comment]
Agree with the escalation. This issue should be a duplicate of #73

**Comment by hrishibhat** [comment]
Result:
Medium
Duplicate of #73

**Comment by sherlock-admin** [comment]
Escalations have been resolved successfully!

Escalation status:
- [0xruhum](https://github.com/sherlock-audit/2023-04-jojo-judging/issues/246/#issuecomment-1571580496): accepted

**Comment by IAm0x52** [comment]
Same fix (and issue) as #74
